lib/vrl/compiler/src/compiler.rs

```diff
@@ -91,44 +88,25 @@ impl<'a> Compiler<'a> {
         Ok((Program { expressions, info }, warnings.into()))
     }
 
-    fn compile_root_exprs(
-        &mut self,
-        nodes: impl IntoIterator<Item = Node<ast::RootExpr>>,
-        external: &mut ExternalEnv,
-    ) -> Vec<Expr> {
-        use ast::RootExpr::*;
-
-        nodes
-            .into_iter()
-            .filter_map(|node| match node.into_inner() {
-                Expr(expr) => {
-                    self.fallible_expression_error = None;
-
-                    let expr = self.compile_expr(expr, external)?;
-
-                    if let Some(error) = self.fallible_expression_error.take() {
-                        self.diagnostics.push(error);
-                    }
-
-                    Some(expr)
-                }
-                Error(err) => {
-                    self.handle_parser_error(err);
-                    None
-                }
-            })
-            .collect()
-    }
-
     fn compile_exprs(
         &mut self,
         nodes: impl IntoIterator<Item = Node<ast::Expr>>,
         external: &mut ExternalEnv,
     ) -> Option<Vec<Expr>> {
-        nodes
-            .into_iter()
-            .map(|node| self.compile_expr(node, external))
-            .collect::<Option<_>>()
+        let mut exprs = vec![];
+        for node in nodes {
+            let expr = self.compile_expr(node, external)?;
+            let type_def = expr.type_def((&self.local, external));
+            exprs.push(expr);
+
+            if type_def.is_never() {
+                // This is a terminal expression. Further expressions must not be
+                // compiled since they will never execute, but could alter the types of
+                // variables in local or external scopes through assignments.
+                break;
+            }
```

Comment:
Do we not hit that path when we compile a branch that might `abort` based on a runtime value?

Reply:
If you `abort` based off of runtime behavior, that means there is a branch somewhere executing 2 different code blocks. This is compiling just a single code block, where a branch isn't allowed.

Reply:
This might not have fully answered your question. You can have an expression inside a code block that _contains_ a branch, and that could conditionally abort. From the perspective of this block that is unconditional though, and the type def of that entire branch has already been correctly calculated. If all branches abort, the type will be `never`, and it knows it unconditionally aborts here. If it conditionally branches, the type is something more concrete than `never`.

Reply:
If we calculated types from assignments through the `type_def` function this wouldn't have been required, but that is a very large refactor at this point.